embassy-nrf/src/usb.rs

```diff
@@ -39,13 +40,37 @@ impl<'d, T: Instance> Driver<'d, T> {
         irq.unpend();
         irq.enable();
 
+        // Initialize the bus so that it signals that power is available.
+        // Not required when using with_power_management as we then rely on the irq.
+        POWER_AVAILABLE.store(true, Ordering::Relaxed);
+        BUS_WAKER.wake();
+
         Self {
             phantom: PhantomData,
             alloc_in: Allocator::new(),
             alloc_out: Allocator::new(),
         }
     }
 
+    /// Establish a new device that then uses the POWER peripheral to
+    /// detect USB power detected/removed events are handled.
+    #[cfg(not(feature = "_nrf5340-app"))]
+    pub fn with_power_management(
```

Comment:
If we're incorporating power management into `embassy-usb`, then this functionality probably needs to be part of the general `new` method. To handle things like the `nrf-softdevice` situation, instead of taking the `power_irq` directly, it should take a generic type which implements a `UsbPowerDetect` trait. That trait can be implemented on a struct which takes an `impl Unborrow<Target = POWER>` for users that can make use of the POWER peripheral directly. Other users can implement their own version of the trait.

Reply:
This is the nrf driver's new we're talking about so stm32 doesn't apply. What I'm saying is that `with_power_management` should replace the nrf driver's `new` method. However, instead of taking the irq directly, it should take a trait to cover the following cases:
* The `POWER` peripheral and/or irq need to be shared with other parts of the application
* `nrf_softdevice` users who don't have access to the `POWER` peripheral
 
My understanding of what @Dirbaio was looking for is to incorporate the Powered state of the USB device state spec into embassy-usb itself. Once we do that it becomes a requirement of the drivers to be able to provide those state transitions to the library. There may be cases where the driver never generates those transitions (e.g. an stm32 device with no IO pin wired to detect Vbus) but the assumption should be that the capability is there.

Reply:
My understanding was that we have events representing power conditions. I guess I'm then thinking that there are sub-states to the existing states given the presence of the power_available bool.

Reply:
I'm still not convinced about the naming/organization of the driver constructor beyond what I've got. It feels ok to me. Perhaps we can have another PR improving on that to what you and/or Dario are thinking about? Thanks.

Reply:
Actually, I saw that I had a bug re. the initial vbus state, so I've fixed that and also provided a means to set power state per softdevice usage .Please see commit 81796d29b44f1b052d7631bc8677525d5aa70b50